packages/medusa/src/services/tax-provider.ts

```diff
@@ -0,0 +1,201 @@
+import { BaseService } from "medusa-interfaces"
+import { EntityManager } from "typeorm"
+import Redis from "ioredis"
+
+import { LineItemTaxLineRepository } from "../repositories/line-item-tax-line"
+import { LineItemTaxLine } from "../models/line-item-tax-line"
+import { Region } from "../models/region"
+import { Cart } from "../models/cart"
+import { Order } from "../models/order"
+import {
+  ITaxService,
+  TaxCalculationContext,
+  TaxCalculationLine,
+} from "../interfaces/tax-service"
+
+import { TaxServiceRate } from "../types/tax-service"
+
+import ProductTaxRateService from "./product-tax-rate"
+
+const CACHE_TIME = 30 // seconds
+
+/**
+ * Provides layer to manipulate users.
+ * @extends BaseService
+ */
+class TaxProviderService extends BaseService {
+  private container_: AwilixContainer
+  private manager_: EntityManager
+  private transactionManager_: EntityManager
+  private productTaxRateService_: ProductTaxRateService
+  private taxLineRepo_: typeof LineItemTaxLineRepository
+  private redis_: Redis
+
+  constructor(container: AwilixContainer) {
+    super()
+
+    this.container_ = container
+    this.taxLineRepo_ = container["lineItemTaxLineRepository"]
+    this.productTaxRateService_ = container["productTaxRateService"]
+    this.eventBus_ = container["eventBusService"]
+    this.manager_ = container["manager"]
+    this.redis_ = container["redisClient"]
+  }
+
+  withTransaction(transactionManager: EntityManager): TaxProviderService {
+    if (!transactionManager) {
+      return this
+    }
+
+    const cloned = new TaxProviderService(this.container_)
+
+    cloned.transactionManager_ = transactionManager
+    return cloned
+  }
+
+  retrieveProvider(region: Region): ITaxService {
+    let provider: ITaxService
+    if (region.tax_provider_id) {
+      provider = this.container_[`tp_${region.tax_provider_id}`] as ITaxService
+    } else {
+      provider = this.container_["systemTaxService"] as ITaxService
+    }
+
+    if (!provider) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Could not find a tax provider with id: ${region.tax_provider_id}`
+      )
+    }
+
+    return provider
+  }
+
+  async getTaxLines(
+    order: Cart | Order,
+    calculationContext: TaxCalculationContext
+  ): Promise<LineItemTaxLine[]> {
+    const calculationLines: TaxCalculationLine[] = await Promise.all(
+      order.items.map(async (l) => {
+        return {
+          item: l,
+          rates: await this.getRegionRatesForProduct(
+            l.variant.product_id,
+            calculationContext.region
+          ),
+        }
+      })
+    )
+
+    const taxProvider = this.retrieveProvider(calculationContext.region)
+    const providerLines = await taxProvider.getTaxLines(
+      calculationLines,
+      calculationContext
+    )
+
+    const liTaxLineRepo = this.manager_.getCustomRepository(this.taxLineRepo_)
+
+    return providerLines.map((pl) => {
+      return liTaxLineRepo.create({
```

Comment:
tiny nit: for people not super familiar with typeorm and coming from other ORMs, could be nice to add a little comment saying something like:
```javascript
return providerLines.map((pl) => {
  // does not save to db
  return liTaxLineRepo.create({
    item_id: pl.item_id,
    rate: pl.rate,
    name: pl.name,
    code: pl.code,
    metadata: pl.metadata,
  })
})
```
Personally had to google it to find out :sweat_smile: